docs/proposals/ingress-ha-scheduling-provisioning.md

```diff
@@ -0,0 +1,194 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.3/docs/proposals/job.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress HA, Scheduling, and Provisioning Proposal
+----
+
+
+## Overview
+Ingress can be used to expose a service in the kubernetes cluster:
+
+* usually cluster admin deploys one Ingress Pod
+* user creates a Ingress resource
+* the Ingress Pod will list&watch All Ingress Resources in the cluster
+* user out of cluster then can access service in the cluster by accessing 
+  the node's ip on which Ingress Pod is running, Ingress Pod will forward 
+  request into cluster based on rules defined in Ingress Resource
+
+This just works. What's the issus then?
+
+The issues is:
+
+* It does not provide High Availability because client needs to know 
+  the IP addresss of the node where Ingress Pod is running. In case of a 
+  failure the Ingress Pod can be be moved to a different node.
+* How many Ingress Pod should run in a cluster? Should all Ingress Pod 
+  list&watch all Ingress Resource with out distinction? There is no way 
+  to bind or schedule ingress resource to a Ingress Pod/ReplicaSet(or a 
+  set of Ingress Pods/ReplicaSets), result in insufficient or excessive 
+  use of resource.
+
+## Goal
+This Proposal aims to address the above issues by the following mechanism:
+
+* Ingress HA: using keepalived and VIP to provide High Availability(mainly
+  for nginx/haproxy implementation, cloud implementation usually already 
+  provide HA)
+* Ingress Scheduling: schedule Ingress Resource to Ingress Pod/ReplicaSet
+* Ingress Provisioning: allow user to dynamically add Ingress Pod/ReplicaSet 
+  on demand
+
+## NonGoal
+
+* Ingress ReplicaSet rolling update
+
+## Ingress HA
+(AKA: Ingress Virtual IP using keepalived)
+
+#### High level design
+* use keepalived to provide HA
+* cluster admin choose a group of nodes which could be accessed out of cluster 
+  and are in the same L2 broacast domain to run Ingress Pod
+* deploy Ingress Pod using ReplicaSet(at least 2 replicas for HA)
+* using AntiAffinity feature so that Ingress Pod created by the same Ingress 
+  ReplicaSet could be scheduled to different node
+* cluster admin choose a CIDR for Ingress VIP(AKA IngreeVIPCIDR)
+* each Ingress Replicaset will be allocated a VIP from IngreeVIPCIDR(allocated by
```

Comment:
In the claims proposal, each ingress claim would get a vip. 
"Ingress Replicaset" is a term which doesn't make sense to me, today an Ingress points to Services which *may* point to replicassets.